### PR TITLE
Enable Online Encryption for data disks.

### PR DESCRIPTION
**Pull request by vimish**
Online encryption of data disk flow:
1. Unmount data volume and create encryption layer.
2. Update crypttab/fstab for the disk
3. Resume parallel encryption for the disks.

This PR is not on our dev branch.
The changes will be merged to dev branch when all the sub features are ready.

### CODE REVIEW

File: VMEncryption/main/handle.py [hunk 1]
```diff
@@ -886,7 +890,8 @@ def enable_encryption():
                 logger.log(msg="config file exists and passphrase file exists.", level=CommonVariables.WarningLevel)
                 encryption_marker = mark_encryption(command=extension_parameter.command,
                                                     volume_type=extension_parameter.VolumeType,
-                                                    disk_format_query=extension_parameter.DiskFormatQuery)
+                                                    disk_format_query=extension_parameter.DiskFormatQuery,
+                                                    encryption_mode='Online' if (disk_util.get_luks_header_size() == CommonVariables.luks_header_size_v2) else None)
```
Comment (vimish) [inline VMEncryption/main/handle.py:894]: To start with keep it only for RHEL.

File: VMEncryption/main/handle.py [hunk 2]
```diff
@@ -1676,6 +1682,14 @@ def enable_encryption_all_in_place(passphrase_file, encryption_marker, disk_util
                            status=CommonVariables.extension_success_status,
                            status_code=str(CommonVariables.success),
                            message=msg)
+    if encryption_marker.get_encryption_mode() == 'Online':
+        logger.log("Starting Online Encryption for data disks.")
+        online_enc_handle = OnlineEncryptionHandler(logger)
+        failed_item = online_enc_handle.handle(device_items_to_encrypt, passphrase_file, disk_util, crypt_mount_config_util, bek_util)
+        if failed_item is not None:
+            return failed_item
+        online_enc_handle.handle_resume_encryption(disk_util)
```
Comment (vimish) [inline VMEncryption/main/handle.py:1691]: Test scenario where some disks are already encrypted.

File: VMEncryption/main/OnlineEncryptionHandler.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            
+            device_dev_path = os.path.join('/dev/', device_item.name)
+            luks_header_size = disk_util.get_luks_header_size()
+            size_shrink_to = (device_item.size - luks_header_size) / CommonVariables.sector_size
+            chk_shrink_result = disk_util.check_shrink_fs(dev_path=device_dev_path, size_shrink_to=size_shrink_to)
+
+            if chk_shrink_result != CommonVariables.process_success:
+                self.logger.log(msg="check shrink fs failed with code {0} for {1}".format(chk_shrink_result, original_dev_path),
+                           level=CommonVariables.ErrorLevel)
+                self.logger.log(msg="your file system may not have enough space to do the encryption.")
```
Comment (vimish) [inline VMEncryption/main/OnlineEncryptionHandler.py:43]: add to Error msg to "FS may not support resizing"

File: VMEncryption/main/OnlineEncryptionHandler.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                return device_item
+
+            mapper_name = str(uuid.uuid4())
+            init_status_code = self.command_executor.ExecuteInBash('cryptsetup reencrypt --encrypt --init-only --reduce-device-size 32768S {1} {2} -d {0} -q'.format(passphrase_file,
```
Comment (vimish) [inline VMEncryption/main/OnlineEncryptionHandler.py:47]: Add reduce device size as constant

File: VMEncryption/main/handle.py [hunk 5]
```diff
@@ -1676,6 +1682,14 @@ def enable_encryption_all_in_place(passphrase_file, encryption_marker, disk_util
                            status=CommonVariables.extension_success_status,
                            status_code=str(CommonVariables.success),
                            message=msg)
+    if encryption_marker.get_encryption_mode() == 'Online':
+        logger.log("Starting Online Encryption for data disks.")
+        online_enc_handle = OnlineEncryptionHandler(logger)
+        failed_item = online_enc_handle.handle(device_items_to_encrypt, passphrase_file, disk_util, crypt_mount_config_util, bek_util)
+        if failed_item is not None:
```
Comment (vimish) [inline VMEncryption/main/handle.py:1689]: block any new operation

**Review by ejarvi — APPROVED** [review]
LGTM

**Pull request merged**